clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerInterceptor.java

```diff
@@ -0,0 +1,65 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license agreements. See the NOTICE
+ * file distributed with this work for additional information regarding copyright ownership. The ASF licenses this file
+ * to You under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the
+ * License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package org.apache.kafka.clients.consumer;
+
+
+import org.apache.kafka.common.Configurable;
+import org.apache.kafka.common.TopicPartition;
+
+import java.util.Map;
+
+/**
+ * A plugin interface to allow things to listen for events happening to the record at different points on the consumer.
+ * <p>
+ * This class will get consumer config properties via <code>configure()</code> method, including clientId assigned
+ * by KafkaConsumer if not specified in the consumer config. The interceptor implementation needs to be aware that it will be
+ * sharing consumer config namespace with other interceptors and serializers, and ensure that there are no conflicts.
+ * <p>
+ * ConsumerInterceptor callbacks are called from the same thread.
+ */
+public interface ConsumerInterceptor<K, V> extends Configurable {
+
+    /**
+     * This is called when the records are about to be returned to the user.
+     * <p>
+     * This method is allowed to mutate consumer records, in which case the new records will be returned.
+     * Any exception thrown by this method will be caught by the caller, logged, but not propagated to the client.
+     * <p>
+     * Since the consumer may run multiple interceptors, a particular interceptor's onConsume() callback will be called in order specified in
+     * configuration {@link org.apache.kafka.clients.consumer.ConsumerConfig#INTERCEPTOR_CLASSES_CONFIG}. The first interceptor
+     * in the list gets the consumed records. The following interceptor will be passed the records returned by the
+     * previous interceptor, and so on. Since interceptors are allowed to mutate records, interceptors may potentially get
+     * the records already mutated by other interceptors. However, building a pipeline of mutable interceptors that depend on the output
```

Comment:
Tthinking about it some more, if the key or value are mutable, then the interceptor could actually mutate those directly. So, maybe it's not wrong. But I think I prefer modify as you suggested. It may be worth getting some other opinions before changing it though.